Add option tls_insecure to allow for insecure TLS

dagwieers opened the pull request:
##### SUMMARY
Some devices only support up to TLSv1.0, resulting now in TLSV1_ALERT_DECODE_ERROR.

Currently module_utils/urls.py already supports TLSv1.0 if the python version does not support anything newer or HAS_SSLCONTEXT is False. Also according to documentation PROTOCOL_SSLv23 should already support PROTOCOL_TLSv1, however that does not seem to be the case (in python 2.7.5 on RHEL7.4): https://docs.python.org/2/library/ssl.html#socket-creation

This PR includes TLSv1.0 support even if python supports newer TLSv1.1 and TLSv1.2, simply because otherwise Ansible cannot manage these older devices.

This could be backported.

PS If needed, we could add TLSv1 support also to other snippets that create an SSL context. In this PR I only considered the open_url() case.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
module_utils/urls.py

##### ANSIBLE VERSION
v2.5

dagwieers commented:
cc @rsmeyers

abadger commented:
We intentionally stopped supporting insecure protocols.  We'll have to ask
for input from the other core maintainers to reverse that decision.  I can
look at whether this code is okay after we come to a consensus about
whether to reverse that decision or not.  (Adding it to the well agenda is
probably the best way to get it talked about)

On Jan 29, 2018 7:00 AM, "Dag Wieers" <notifications@github.com> wrote:

> @dagwieers <https://github.com/dagwieers> requested your review on:
> ansible/ansible#35462 <https://github.com/ansible/ansible/pull/35462>
> Support TLSv1 with open_url().
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/ansible/ansible/pull/35462#event-1446490148>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAMxWm8Be7EBKD5GPFDEpRq38ktAo26kks5tPd0bgaJpZM4Rwu4l>
> .
>

dagwieers commented:
@abadger I am all for that by default, but I think we should still support it. We are getting reports from both HP iLO users as well as Cisco IMC that we need a way to support older devices (even just to be able to upgrade them to newer firmware).

A second use-case for current $CUSTOMER we still need to support older firmware/software/devices, so we require automated testing that works even if newer firmware/releases exist that support newer SSL/TLS versions.

bcoca commented:
So i would not enable this by default, but force users to 'unsecure' themselves via a configuration option in which the explicitly enable the protocols

dagwieers committed:
Add option tls_insecure to allow for insecure TLS

Some devices only support up to TLSv1.0, resulting now in
TLSV1_ALERT_DECODE_ERROR.

Currently module_utils/urls.py already supports TLSv1.0 if the python
version does not support anything newer or HAS_SSLCONTEXT is False. Also
according to documentation PROTOCOL_SSLv23 should already support
PROTOCOL_TLSv1, however that does not seem to be the case (in python
2.7.5 on RHEL7.4):
https://docs.python.org/2/library/ssl.html#socket-creation

This PR includes TLSv1.0 support even if python supports newer TLSv1.1
and TLSv1.2, simply because otherwise Ansible cannot manage these older
devices.

dagwieers commented:
@abadger @bcoca I added a boolean option `tls_insecure` to allow for older SSL/TLS methods.

dagwieers committed:
Support new tls_insecure option to modules

dagwieers committed:
Fix unit tests

dagwieers committed:
Changes request during review

dagwieers commented:
Please re-review !

felixfontein commented on lib/ansible/modules/remote_management/imc/imc_rest.py:
```diff
@@ -296,10 +303,10 @@ def imc_response(module, rawoutput, rawinput=''):
     ''' Perform a logout, if needed '''
     data = '<aaaLogout cookie="%s" inCookie="%s"/>' % (cookie, cookie)
-    resp, auth = fetch_url(module, url, data=data, method="POST", timeout=timeout)
+    resp, auth = fetch_url(module, url, data=data, method="POST", timeout=timeout, tls_insecure=tls_insecure)
```
`fetch_url` does not have a `tls_insecure` parameter; it takes the value from `module.params['tls_insecure']` instead.

dagwieers replied:
Good catch, fixed.

felixfontein commented on lib/ansible/utils/module_docs_fragments/url.py:
```diff
@@ -61,4 +60,11 @@ class ModuleDocFragment(object):
+  tls_insecure:
+    description:
+      - Enable insecure TLS protocols.
+      - This may be required to support older devices.
```
I think this (and the other module docs for `tls_insecure`) should contain a warning that this option should only be used if really needed, because it is insecure. (Cannot be mentioned often enough. Otherwise it might end up as a "magic solution" similarly to setting `validate_certs: no` everywhere, which "solves" a lot of problems and introduces a lot of security holes.)

dagwieers replied:
I added a statement similar to what is used for `validate_certs`. However, enabling `tls_insecure` doesn't disable secure TLS options, it simply allows for falling back to insecure (older) TLS options if more secure alternatives are not available.

sivel commented:
I believe `tls_insecure` will need to be exposed to `SSLValidationHandler` and `CustomHTTPSConnection` as well.  When doing so, we need to validate that the method used works for `PyOpenSSLContext` also.

dagwieers commented:
@sivel Can you re-review this PR ?

sivel commented on lib/ansible/module_utils/urls.py:
```diff
@@ -308,6 +309,11 @@ def __init__(self, *args, **kwargs):
+
+            # Support insecure TLS protocol options
+            # if self.context and self.tls_insecure:
+            #     tls_insecure_options(self.context)
```
It appears to support this it will require adding an `__init__` method to `CustomHTTPSHandler` so that `tls_insecure` can be stored and passed to `CustomHTTPSConnection`.  See `HTTPSClientAuthHandler` for an example of what would be required.

dagwieers replied:
This is now implemented.

sivel commented on lib/ansible/module_utils/urls.py:
```diff
@@ -385,6 +391,21 @@ def as_list(self):
+    '''
+    # TODO: We may want to enable older ciphers here too ?
+    # context.set_ciphers('HIGH:!aNULL:!eNULL')
+    context.options &= ~ssl.OP_NO_SSLv2
```
We need to check if `ssl.OP_NO_SSLv2` it not `None` before using it.  Newer versions of openssl have removed it, which may cause issues here. See https://github.com/ansible/ansible/commit/496d10f

dagwieers replied:
Makes sense, but then I prefer to compare to None explicitly. Thanks!

dagwieers commented:
@sivel Please re-review!

puiterwijk commented on lib/ansible/modules/monitoring/grafana_dashboard.py:
```diff
@@ -100,6 +100,13 @@
     version_added: 2.7
+  tls_insecure:
+    description:
+      - Enable insecure TLS protocols.
```
Perhaps change this to `Enable insecure SSL protocols.`?
Since what we're enabling here is actually SSL, not TLS.
Or alternatively, explicitly mention that it's enabling SSLv2 and SSLv3?

felixfontein replied:
It also enables TLSv1 (if Python supports a newer TLS version) according to the PR description. So `Enable insecure TLS/SSL protocols.` would maybe be a better description?

puiterwijk replied:
That'd work, yeah.

mattclay commented:
CI failure in unit tests: https://app.shippable.com/github/ansible/ansible/runs/111089/7/tests

dagwieers committed:
Fix CI issue and improve description as suggested

dagwieers commented:
@sivel Please review

sivel commented on lib/ansible/module_utils/urls.py:
```diff
@@ -317,12 +318,21 @@ class NoSSLError(SSLValidationError):
 if hasattr(httplib, 'HTTPSConnection') and hasattr(urllib_request, 'HTTPSHandler'):
     class CustomHTTPSConnection(httplib.HTTPSConnection):
         def __init__(self, *args, **kwargs):
+            self.tls_insecure = kwargs['tls_insecure']
```
Use `kwargs.pop` here instead of this and a `del` later.

sivel commented on lib/ansible/module_utils/urls.py:
```diff
@@ -917,7 +955,8 @@ def rfc2822_date_string(timetuple, zone='-0000'):
                  url_username=None, url_password=None, http_agent=None, force_basic_auth=False,
-                 follow_redirects='urllib2', client_cert=None, client_key=None, cookies=None, unix_socket=None):
+                 follow_redirects='urllib2', client_cert=None, client_key=None, cookies=None,
+                 unix_socket=None, tls_insecure=False):
```
We need a `self. tls_insecure = tls_insecure` in this method.

sivel commented on lib/ansible/module_utils/urls.py:
```diff
@@ -966,7 +1005,7 @@ def open(self, method, url, data=None, headers=None, use_proxy=None,
              force_basic_auth=None, follow_redirects=None,
              client_cert=None, client_key=None, cookies=None, use_gssapi=False,
-             unix_socket=None):
+             unix_socket=None, tls_insecure=False):
```
This should default to `None` here.  And further below we need:

```
tls_insecure = self._fallback(tls_insecure, self. tls_insecure)
```

sivel replied:
This is also missing a docstring description of the new argument below.

bcoca commented:
closing due to inactivity, feel free to reopen PR if this is something you still want to pursue.
